include/rocksdb/utilities/cache_dump_load.h

```diff
@@ -0,0 +1,77 @@
+//  Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+#include <set>
+
+#include "rocksdb/cache.h"
+#include "rocksdb/env.h"
+#include "rocksdb/io_status.h"
+#include "rocksdb/secondary_cache.h"
+#include "rocksdb/table.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+class CacheDumpWriter {
+ public:
+  virtual ~CacheDumpWriter() = default;
+
+  virtual IOStatus Write(const Slice& data) = 0;
+  virtual IOStatus Close() = 0;
+  virtual uint64_t GetFileSize() = 0;
+};
+
+class CacheDumpReader {
+ public:
+  virtual ~CacheDumpReader() = default;
+  virtual IOStatus Read(size_t len, std::string* data) = 0;
+  virtual size_t GetOffset() const = 0;
+  virtual IOStatus Close() = 0;
+};
+
+// CacheDumpOptions is used to control DumpCache()
+struct CacheDumpOptions {
+  // The env pointer
+  Env* env;
+  // The prefix set, which filter out the cache keys that does not match
+  std::set<std::string> prefix_set;
+};
+
+class CacheDumper {
+ public:
+  virtual ~CacheDumper() = default;
+  virtual Status Prepare() = 0;
+  virtual IOStatus Run() = 0;
+};
+
+class CacheDumpedLoader {
+ public:
+  virtual ~CacheDumpedLoader() = default;
+  virtual Status Prepare() = 0;
+  virtual IOStatus Run() = 0;
+};
+
+Status NewDefaultCacheDumpWriter(Env* env, const EnvOptions& env_options,
+                                 const std::string& file_name,
+                                 std::unique_ptr<CacheDumpWriter>* writer);
+
+Status NewDefaultCacheDumpReader(Env* env, const EnvOptions& env_options,
+                                 const std::string& file_name,
+                                 std::unique_ptr<CacheDumpReader>* reader);
+
+Status NewDefaultCacheDumper(const CacheDumpOptions& dump_options,
+                             const std::shared_ptr<Cache>& cache,
+                             std::unique_ptr<CacheDumpWriter>&& writer,
+                             std::unique_ptr<CacheDumper>* cache_dumper);
+
+Status NewDefaultCacheDumpedLoader(
+    const CacheDumpOptions& dump_options,
+    const BlockBasedTableOptions& toptions,
+    const std::shared_ptr<SecondaryCache>& secondary_cache,
```

Comment:
1. For now, it is easier to load to the secondary cache. But, yes, I think we should leave the space for extension in the future to load to the block cache.
2. You mean, directly add the "dump" and "load" to the Cache and SecondaryCache class?

Reply:
Yes, I mean why not have a method off the Cache/Secondary that do Dump/Load.  Why introduce a new mechanism to iterate over the cache when the cache could do it itself.

Reply:
> Yes, I mean why not have a method off the Cache/Secondary that do Dump/Load. Why introduce a new mechanism to iterate over the cache when the cache could do it itself.

My take is, in the future, the dump and load function can be more complicated: dump and load step-by-step, more complicated filtering, control the dumping and loading rate and etc. It might be better to have a new mechanism for this process instead of add them to the cache. Just leave the cache classes simple

Reply:
While I agree that there might be additions to this model in the future, I think it is more likely that the CacheReader/Writer interfaces need extended.  For example, if you wanted to do rate limiting, I believe the CacheWriter would need to expose an interface.  Otherwise, the Cache may be blocked attempting to interact with a Writer while the Cache holds a lock.  The alternatives are that you must be able to iterate over the Cache in a "lock free" means or that the Loader needs to know how much it can accept from the Cache in a given block.

The point is that I think there are more complications to these sort of additional features and that simply adding a class does not make them any simpler or eliminate them.

Reply:
In general, the dumper and loader is more like a independent tool from my point of view......